Add backend for conversation descriptions

danxuliu opened the pull request:
Requires #4656
Probably should be merged after #4496 (there will be conflicts, but it should be easier to fix them here)

Fixes (the backend part of) #3432

Only moderators and guest moderators of group and public conversations can set their description.

Setting and getting the description is available only in version 3 of the API.

The description is returned along the rest of the information of a conversation. ~~However, as the description is not typically needed in the conversation list, it could be potentially "large" and the conversation list is fetched frequently only a hash of the description is returned in that case. The exception is when the description is empty, in which case an empty value is returned, instead of the value of `sha1('')`.~~

~~If only a single conversation is fetched the actual description is always returned.~~ Due to design reasons the description is now limited to just 250 characters. Moreover, the OCS endpoint is gzip compressed, so sending the raw description instead of the hash when fetching the conversation list does not reduce the size too much, but introduces an inconsistency between the data of a single conversation and all conversations, which requires a more complex logic in the clients (and specifically in the WebUI it would require quite some changes). Due to all this, for simplicity the raw description is also returned when getting the conversation list.

Also for simplicity the owner of a password request conversation can currently set the description, even if there is probably no point on doing that. Preventing it would be easy, though, just something similar to https://github.com/nextcloud/spreed/pull/4656/commits/630e2f4acb89b90c6ba718d240b4f0a92dce317a

When the description is changed the external signaling server is notified and the new description is sent; the description, like the other properties, is sent whenever the conversation is modified, even if the description itself was not modified, but (if I recall correctly) this is needed for the external signaling server to track the changes.

Pending:
- [X] Return description too in [SIP bridge requests](https://github.com/nextcloud/spreed/blob/6cc98c58ab12773bbcc2421488473238d6b68b44/lib/Controller/RoomController.php#L559-L572)? @nickvergessen -> Not needed
- [X] Handle error setting a description too long in the database? - The field is now declared as TEXT without a limit, so in practice the description set will never be too long
- [ ] How to check that the database supports multibyte strings to use `strlen` instead of `mb_strlen` if it does not? Or remove the guard from the controller and just handle if there is an error when setting the description in the database? Or do both (for example, always guard with `mb_strlen`, but also handle the error)?
- [X] Add integration tests for video verification and file rooms
- [X] Add support for conversation description to occ commands
- [X] Send system message when the description is changed
- [X] Add documentation
- Something else? Not sure if I am missing something

nickvergessen commented:
I don't think the SIP bridge needs the description. If you put a link in there e.g. for a shared presentation, people would get it read out or what?🙈

danxuliu commented:
> I don't think the SIP bridge needs the description. If you put a link in there e.g. for a shared presentation, people would get it read out or what?see_no_evil

I have no idea, that is why I asked :-P But ok, no description for SIP bridge then ;-)

nickvergessen commented on docs/conversation.md:
```diff
@@ -29,6 +29,7 @@
         `type` | int | * | See list of conversation types in the [constants list](constants.md#Conversation-types)
         `name` | string | * | Name of the conversation (can also be empty)
         `displayName` | string | * | `name` if non empty, otherwise it falls back to a list of participants
+        'description` | string | v3 | Description of the conversation (can also be empty)
```
```suggestion
        `description` | string | v3 | Description of the conversation (can also be empty)
```

nickvergessen commented on docs/conversation.md:
```diff
@@ -129,6 +130,25 @@
+* Response:
+    - Status code:
+        + `200 OK`
+        + `400 Bad Request` When the name is too long
```
*descriptino

nickvergessen commented on lib/Chat/Parser/SystemMessage.php:
```diff
@@ -119,6 +119,20 @@ public function parseMessage(Message $chatMessage): void {
 				$parsedMessage = $this->l->t('An administrator renamed the conversation from "%1$s" to "%2$s"', [$parameters['oldName'], $parameters['newName']]);
 			}
+		} elseif ($message === 'description_set') {
+			$parsedMessage = $this->l->t('{actor} set the description to "%1$s"', [$parameters['newDescription']]);
```
I would not put the description into the system message.
It can be potentially long and also contain `"` and then look broken.

Whatsapp and others just write "{actor} updated the description" alike messages.

nickvergessen replied:
And we can then use this for create, update and remove at the same time.

danxuliu replied:
@ma12-co checked other platforms and asked for the description to be shown in the message.

ma12-co replied:
I think it should be shown, given that:
1) It's likely to be very important information for the conversation participants;
2) It's s going to be relatively concise since we're capping the max length to 500 chars, so it's not going to be disruptive;

nickvergessen commented on lib/Chat/SystemMessage/Listener.php:
```diff
@@ -122,6 +122,19 @@ public static function register(IEventDispatcher $dispatcher): void {
+		$dispatcher->addListener(Room::EVENT_AFTER_DESCRIPTION_SET, static function (ModifyRoomEvent $event) {
+			$room = $event->getRoom();
+			/** @var self $listener */
+			$listener = \OC::$server->query(self::class);
```
query is deprecated, use
```suggestion
			$listener = \OC::$server->get(self::class);
```

nickvergessen commented on lib/Controller/RoomController.php:
```diff
@@ -971,6 +973,29 @@ public function renameRoom(string $roomName): DataResponse {
+			return new DataResponse([], Http::STATUS_BAD_REQUEST);
+		}
+
+		$description = trim($description);
```
trim before checking the length like in `lib/Command/Room/TRoomCommand.php`

nickvergessen replied:
Also all this logic could be in `$this->room->setDescription` so you only need to do it in one place

nickvergessen commented on lib/Migration/Version11000Date20201011082810.php:
```diff
@@ -0,0 +1,66 @@
+
+class Version11000Date20201011082810 extends SimpleMigrationStep {
+
+	/** @var IDBConnection */
+	protected $connection;
+
+	public function __construct(IDBConnection $connection) {
+		$this->connection = $connection;
+	}
```
```suggestion
```

nickvergessen commented on lib/Migration/Version11000Date20201011082810.php:
```diff
@@ -0,0 +1,66 @@
+			]);
+		}
+
+		return $schema;
```
Put this inside `if (!$table->hasColumn('description')) {`

and return null otherwise. Speeds things up a bit in that case (saving ~20sec on oracle)

danxuliu replied:
Note that I just copy-pasted from other migrations, so I guess that they should be fixed as well (but in another pull request, of course :-) ).

nickvergessen replied:
yeah, just because we did things in the past, doesnt mean we should not improve with new code :)

nickvergessen commented on lib/Room.php:
```diff
@@ -97,6 +99,8 @@ class Room {
 	public const EVENT_AFTER_SESSION_LEAVE_CALL = self::class . '::postSessionLeaveCall';
 	public const EVENT_BEFORE_SIGNALING_PROPERTIES = self::class . '::beforeSignalingProperties';
 
+	public const DESCRIPTION_MAXIMUM_LENGTH = 250;
```
we might need some buffer between table column size and the allowed length.
Also 250 is rather short. If you want to put in a link to a calendar, github project board and some more info you are over it already.

I would say the column size should be 4k and the allowed size 2k or something? Or we even go further and just use `text` and limit it to 32k with this similar to chat messages.

nickvergessen replied:
cc @jancborchardt

danxuliu replied:
> we might need some buffer between table column size and the allowed length.

You are the database master, so just let me know what I have to put in the migration :-P

> Also 250 is rather short. 

Initially the length was larger, but @ma12-co said that the front-end will only allow 250 characters due to design reasons.

nickvergessen replied:
Well let's make the database save, so we can adjust it easily and don't have to run migrations (especially because changing from string to text doesnt work in oracle

nickvergessen commented on lib/Room.php:
```diff
@@ -327,6 +340,12 @@ public function getPropertiesForSignaling(string $userId): array {
 		];
 
+		if ($roomModified) {
+			$properties = array_merge($properties, [
+				'description' => $this->getDescription(),
+			]);
```
This is not changing anything on the signaling?

danxuliu replied:
What do you mean?

nickvergessen replied:
This is `getPropertiesForSignaling` but the data is not used in signaling, so we don't need to expose it?

danxuliu replied:
The new description will be relayed by the signaling server to the clients in the `roomlist->update` message. Currently we are not using the updated properties from the message, though, but I think it is better to send it for consistency with the rest of the update messages.

nickvergessen commented on tests/integration/features/command/create.feature:
```diff
@@ -51,3 +51,31 @@ Feature: create
+
+
+
+  Scenario: Create a room public room for participant1 as owner with a description of 200 characters
```
```suggestion
  Scenario: Create a public room for participant1 as owner with a description of 200 characters
```

nickvergessen commented on tests/integration/features/command/create.feature:
```diff
@@ -51,3 +51,31 @@ Feature: create
+      | name  | type | participantType | description |
+      | room1 | 3    | 1               | 012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678C012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678C0123456789012345678901234567890123456789012345678L |
+
+  Scenario: Create a room public room for participant1 as owner with a description of more than 200 characters
```
```suggestion
  Scenario: Create a public room for participant1 as owner with a description of more than 200 characters
```

nickvergessen commented on tests/integration/features/command/create.feature:
```diff
@@ -51,3 +51,31 @@ Feature: create
+    Then the command failed with exit code 1
+    And the command output contains the text "Invalid room description"
+
+  Scenario: Create a room public room for participant1 as owner with a description of 200 multibyte characters
```
```suggestion
  Scenario: Create a public room for participant1 as owner with a description of 200 multibyte characters
```

nickvergessen commented on tests/integration/features/command/create.feature:
```diff
@@ -51,3 +51,31 @@ Feature: create
+      | name  | type | participantType | description |
+      | room1 | 3    | 1               | ०१२३४५६७८९०१२३४५६७८९०१२३४५६७८९०१२३४५६७८९०१२३४५६७८९०१२३४५६७८९०१२३४५६७८९०१२३४५६७८९०१२३४५६७८९०१२३४५६७८च०१२३४५६७८९०१२३४५६७८९०१२३४५६७८९०१२३४५६७८९०१२३४५६७८९०१२३४५६७८९०१२३४५६७८९०१२३४५६७८९०१२३४५६७८९०१२३४५६७८च०१२३४५६७८९०१२३४५६७८९०१२३४५६७८९०१२३४५६७८९०१२३४५६७८ल |
+
+  Scenario: Create a room public room for participant1 as owner with a description of more than 200 multibyte characters
```
```suggestion
  Scenario: Create a public room for participant1 as owner with a description of more than 200 multibyte characters
```

nickvergessen commented on lib/Migration/Version11000Date20201011082810.php:
```diff
@@ -0,0 +1,61 @@
+
+		$table = $schema->getTable('talk_rooms');
+		if (!$table->hasColumn('description')) {
+			$table->addColumn('description', Type::STRING, [
+				'notnull' => false,
+				// Room::DESCRIPTION_MAXIMUM_LENGTH is not used as changes to
+				// the value would require a migration.
+				'length' => 250,
```
```suggestion
			$table->addColumn('description', Type::TEXT, [
				'notnull' => false,
				'default' => '',
```

danxuliu commented:
Squashed and rebased. It should be ready for a final review :-)

The only missing thing would be the _How to check that the database supports multibyte strings to use `strlen` instead of `mb_strlen` if it does not?_. However, given that the database is now a text field with no length set I **guess** that is no longer an issue, as even if the string contains multibyte characters but the database does not support them this would not cause the description to be too long to be stored in the database.

nickvergessen commented:
> The only missing thing would be the How to check that the database supports multibyte strings to use strlen instead of mb_strlen if it does not?. However, given that the database is now a text field with no length set I guess that is no longer an issue, as even if the string contains multibyte characters but the database does not support them this would not cause the description to be too long to be stored in the database.

mb is always better to use, as it prevents cutting inside a utf8 char

nickvergessen commented:
> Conflicting files
tests/integration/features/bootstrap/FeatureContext.php


I will rebase it, happened with #4702

nickvergessen commented:
Simply dropped "Guard against using undefined 'participants' key" which did exactly the same as https://github.com/nextcloud/spreed/pull/4702/commits/6206ca7b9aefd7de09637530ac0f8d44cc3e238c just a few lines further up
